### PR TITLE
Nx compatibility based on making Graph subclass and calling Cugraph algos

### PR DESCRIPTION
**Pull request by acostadon**
Code to inherit all of networkx graph then call cugraph pagerank with just an import change. 
This was originally going to do bfs first but pargerank was considered a more useful first algorithm.

### CODE REVIEW

**Comment by rlratzel** [comment]
Removed "do not merge" label so CI would run when the "[skip-ci]" tag was removed and since Draft PRs can't be merged anyway.

**Comment by rlratzel** [comment]
rerun tests

**Comment by rlratzel** [comment]
raft update published which should resolve thrust cmake issue in CI.

rerun tests

**Comment by BradReesWork** [comment]
rerun tests

File: python/cugraph/cugraph/compat/nx/algorithms/link_analysis/pagerank_alg.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import cugraph
+import cugraph.utilities
+import cudf
+import numpy as np
+
+
+def cudify(d):
+    if d is None:
+        return None
+
```
[the remark this answers is not in the document]
Comment (acostadon) [reply python/cugraph/cugraph/compat/nx/algorithms/link_analysis/pagerank_alg.py]: Done

**Review by BradReesWork — APPROVED** [review]

File: python/cugraph/cugraph/experimental/compat/nx/__init__.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    importlib.import_module("networkx"), __name__)
+
+del _visited
+del _import_submodules_recursively
+
+# At this point, individual types that cugraph.nx are overriding
+# could be used to override the corresponding types *inside* the
+# networkx modules imported above. For example, the networkx graph generators
+# will still return networkx.Graph objects instead of cugraph.nx.Graph
+# objects (unless the userknows to pass a "create_using" arg, if available).
```
Comment (jnke2016) [inline python/cugraph/cugraph/experimental/compat/nx/__init__.py:91]: There is a typo here I think

File: python/cugraph/cugraph/experimental/compat/nx/__init__.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+# For specific overrides, assignments could be made in the imported
```
Comment (jnke2016) [inline python/cugraph/cugraph/experimental/compat/nx/__init__.py:92]: Typo in this sentence

File: python/cugraph/cugraph/experimental/compat/nx/algorithms/link_analysis/pagerank_alg.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+import cugraph
+import cugraph.utilities
+import cudf
+import numpy as np
+
+
+def cudify(d):
+    """
+    Converts a dictionary which is what networkx sends to a
```
Comment (jnke2016) [inline python/cugraph/cugraph/experimental/compat/nx/algorithms/link_analysis/pagerank_alg.py:22]: Does `networkx` only returns dictionary . probably reword it to something like `convert networkx dictionary results to a cudf.Dataframe`. cuGraph doesn't require cudf.DataFrame for all inputs parameters. Some of them can be lists

File: python/cugraph/cugraph/experimental/compat/nx/algorithms/link_analysis/pagerank_alg.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    Returns
+    -------
+    a DataFrame of (vertex)ids and values.
+    """
+    if d is None:
+        return None
+    if not(isinstance(d, dict)):
+        raise TypeError("type_name must be a dict, got: "
+                        f"{type(d)}")
+    k = np.fromiter(d.keys(), dtype="int32")
```
Comment (jnke2016) [inline python/cugraph/cugraph/experimental/compat/nx/algorithms/link_analysis/pagerank_alg.py:38]: This can be a very expensive operation. `cudf.DataFrame` support dictionaries as input so you can directly pass it like this` df = cudf.DataFrame(d)` and rename the column name with the `.rename` function

File: python/cugraph/cugraph/experimental/compat/nx/algorithms/link_analysis/pagerank_alg.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+    Returns
+    -------
+        PageRank : dictionary
+               A dictionary of nodes with the PageRank as value
+
+    """
+    # Purposely leaving in this statement to demonstrate when cugraph
+    # algorith is being called.
+    print("Calling compat.nx pagerank")
```
Comment (jnke2016) [inline python/cugraph/cugraph/experimental/compat/nx/algorithms/link_analysis/pagerank_alg.py:112]: Is this print left here on purpose or is it a debug print?

File: python/cugraph/cugraph/experimental/compat/nx/algorithms/link_analysis/pagerank_alg.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        local_nstart = cudify(nstart)
+    return cugraph.pagerank(
+            G,
+            alpha,
+            local_pers,
+            max_iter,
+            tol,
+            local_nstart,
+            weight,
+            dangling)
```
Comment (jnke2016) [inline python/cugraph/cugraph/experimental/compat/nx/algorithms/link_analysis/pagerank_alg.py:127]: If the `dangling` parameter is ignored. Why are you passing it though? And this would cause an issue

File: python/cugraph/cugraph/tests/test_compat_algo.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+def test_connectivity():
+    expected = list([{1, 2, 3, 4, 5}, {8, 9, 7}])
+    G = nx.Graph()
+    G.add_edges_from([(1, 2), (2, 3), (3, 4), (4, 5)])
+    G.add_edges_from([(7, 8), (8, 9), (7, 9)])
+    print(list(nx.connected_components(G)))
+    assert list(nx.connected_components(G)) == expected
+
+
+def test_pagerank():
```
Comment (jnke2016) [inline python/cugraph/cugraph/tests/test_compat_algo.py:26]: If you are only checking for the results type returned by `nx.pagerank` ,perhabs rename this function to something more explicit like `test_nx_pagerank_result_type`

File: python/cugraph/cugraph/tests/test_compat_algo.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+def test_pagerank_result_type():
+    G = nx.DiGraph()
+    [G.add_node(k) for k in ["A", "B", "C", "D", "E", "F", "G"]]
+    G.add_edges_from([('G', 'A'), ('A', 'G'), ('B', 'A'),
+                     ('C', 'A'), ('A', 'C'), ('A', 'D'),
+                     ('E', 'A'), ('F', 'A'), ('D', 'B'),
+                     ('D', 'F')])
+    ppr1 = nx.pagerank(G)
+    # This just tests that the right type is returned.
+    assert type(ppr1) == dict
```
Comment (jnke2016) [inline python/cugraph/cugraph/tests/test_compat_algo.py:34]: Probably `isinstance` is better

File: python/cugraph/cugraph/tests/test_compat_algo.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cugraph.experimental.compat.nx as nx
+
+
+def test_connectivity():
+    expected = list([{1, 2, 3, 4, 5}, {8, 9, 7}])
```
Comment (jnke2016) [inline python/cugraph/cugraph/tests/test_compat_algo.py:18]: Why are you converting a list that is already a list?
`list([{1, 2, 3, 4, 5}, {8, 9, 7}]) == [{1, 2, 3, 4, 5}, {8, 9, 7}]`

File: python/cugraph/cugraph/tests/test_compat_pr.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+                                               )
+
+
+@pytest.fixture(scope="module", params=fixture_params)
+def input_combo(request):
+    """
+    Simply return the current combination of params as a dictionary for use in
+    tests or other parameterized fixtures.
+    """
+    print("parameters are \n", request.param, flush=True)
```
Comment (jnke2016) [inline python/cugraph/cugraph/tests/test_compat_pr.py:75]: another print statement here. We usually only print mismatches, modules version and benchmark numbers.

File: python/cugraph/cugraph/tests/test_compat_pr.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+    M = utils.read_csv_for_nx(input_combo["graph_file"])
+
+    Gnx = networkx.from_pandas_edgelist(
+        M, source="0", target="1", edge_attr="weight",
+        create_using=networkx.DiGraph()
+    )
+    nnz_vtx = np.unique(M[['0', '1']])
+    personalization = get_personalization(input_combo["pers_percent"],
+                                          nnz_vtx)
+    print(type(Gnx))
```
Comment (jnke2016) [inline python/cugraph/cugraph/tests/test_compat_pr.py:102]: Another print

File: python/cugraph/cugraph/tests/test_compat_algo.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    G = nx.Graph()
+    G.add_edges_from([(1, 2), (2, 3), (3, 4), (4, 5)])
+    G.add_edges_from([(7, 8), (8, 9), (7, 9)])
+    assert list(nx.connected_components(G)) == expected
+
+
+def test_pagerank_result_type():
```
Comment (jnke2016) [inline python/cugraph/cugraph/tests/test_compat_algo.py:25]: You have a separate test file `test_compat_pr.py` where you are testing `pagerank`. Why not add `test_pagerank_result_type` there? is there a reason to have the pagerank tests in different files? what is `test_compat_algo.py` meant  for? to hold light tests for all algos?

File: python/cugraph/cugraph/tests/test_compat_pr.py [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+
+    M = utils.read_csv_for_nx(graph_file)
+    Gnx = nx.from_pandas_edgelist(
+        M, source="0", target="1", edge_attr="weight",
+        create_using=nx.DiGraph()
+    )
+    pr = nx.pagerank(Gnx)
+
+    # Rounding issues show up in runs but this tests that the
+    # cugraph and networkx algrorithmsare being correctly called.
```
Comment (jnke2016) [inline python/cugraph/cugraph/tests/test_compat_pr.py:158]: Typo

**Review by BradReesWork — APPROVED** [review]
